### PR TITLE
Add double support

### PR DESCRIPTION
**Pull request by csarofeen**
Pipe double support through the fusion system
Includes random number generation
Fixes to code generation (floating point scalar ivalue inputs are always double)
Generic suffix support for operators (e.g. fmaxf / sinf)
Testing in .cpp minor testing in .py (.py testing should be improved cc @shmsong, when this gets merged please improve double testing in the python tests)

### CODE REVIEW

**Comment by csarofeen** [comment]
Strangely, flake8 seems unrelated to pr.

File: test/test_jit_cuda_fuser.py [hunk 1]
```diff
@@ -155,16 +155,11 @@ def t(x: torch.Tensor):
             return o
         t_jit = torch.jit.script(t)
 
-        prev_fallback = os.environ['PYTORCH_NVFUSER_DISABLE_FALLBACK']
```
Comment (tlemo) [inline test/test_jit_cuda_fuser.py:158]: this doesn't seem related to double support?

File: test/cpp/jit/test_gpu.cpp [hunk 2]
```diff
@@ -2888,40 +2891,42 @@ IValue gen_aten_operand(
...
+        return IValue(
+            at::rand({blocks, threads}, options).round().to(at::kBool));
       } else {
         auto options =
             at::TensorOptions().dtype(at::kBool).device(at::kCUDA, 0);
         return IValue(at::empty({blocks, threads}, options));
       }
     } else {
-      TORCH_CHECK("Not currently supported type", desc.second)
+      TORCH_CHECK(false, "Not currently supported type: ", desc.second)
```
Comment (tlemo) [inline test/cpp/jit/test_gpu.cpp:2916]: nice catch - maybe we should file an upstream bug to add a linter check for TORCH_CHECK(true, ...) ?

File: torch/csrc/jit/codegen/cuda/codegen.cpp [hunk 3]
```diff
@@ -60,10 +60,42 @@ class CudaKernelGenerator : private kir::IrVisitor {
...
+            break;
+          default:
+            TORCH_INTERNAL_ASSERT(
+                false,
+                "Scalar type of ",
+                val->dtype(),
+                " is not currently supported as a scalar argument to kernels.");
+        }
+        if (isFloatingPoint) {
+          code_ << DataType::Double;
```
Comment (tlemo) [inline torch/csrc/jit/codegen/cuda/codegen.cpp:85]: why use double for all floating point types?

File: torch/csrc/jit/codegen/cuda/codegen.cpp [hunk 4]
```diff
@@ -60,10 +60,42 @@ class CudaKernelGenerator : private kir::IrVisitor {
...
+        } else {
+          code_ << DataType::Int;
+        }
+        if (val->definition() != nullptr) {
```
Comment (tlemo) [inline torch/csrc/jit/codegen/cuda/codegen.cpp:89]: is this check really needed? why not always use gen(val), as before?

File: torch/csrc/jit/codegen/cuda/executor_kernel_arg.h [hunk 5]
```diff
@@ -69,17 +70,9 @@ struct LongArg : public ArgAbstract {
...
-    return &val_;
-  }
-};
-
-struct FloatArg : public ArgAbstract {
-  float val_;
-  FloatArg(float _val) : val_(_val){};
+struct DoubleArg : public ArgAbstract {
+  double val_;
+  DoubleArg(double _val) : val_(_val){};
```
Comment (tlemo) [inline torch/csrc/jit/codegen/cuda/executor_kernel_arg.h:75]: we should add explicit too all these constructors

File: torch/csrc/jit/codegen/cuda/kernel_ir_printer.cpp [hunk 6]
```diff
@@ -229,12 +238,16 @@ void IrPrinter::visit(const kir::UnaryOp* node) {
...
+      if (needFloatSuffix(node->operation()) &&
+          node->out()->dtype() == DataType::Float) {
+        ir_str_ << "f";
+      }
     }
 
-    ir_str_ << "(";
     if (node->operation() == UnaryOpType::RandLike) {
-      ir_str_ << "RND";
+      ir_str_ << "(RND";
```
Comment (tlemo) [inline torch/csrc/jit/codegen/cuda/kernel_ir_printer.cpp:248]: thanks for fixing this

File: torch/csrc/jit/codegen/cuda/type.cpp [hunk 7]
```diff
@@ -89,50 +91,109 @@ static const char* expr_type2string(ExprType t) {
   }
 }
 
+bool needFloatSuffix(UnaryOpType t) {
+  switch (t) {
+    case UnaryOpType::Acos:
```
Comment (tlemo) [inline torch/csrc/jit/codegen/cuda/type.cpp]: wouldn't be easier to list the cases where we _don't_ need the suffix (and return true on the default path) ?

**Review by tlemo — APPROVED** [review]

File: test/test_jit_cuda_fuser.py (hunk 1 above)
Comment (csarofeen) [reply test/test_jit_cuda_fuser.py:158]: It is, as we weren't fusing double operations so we expected the test to go through the fusion fallback. But now we are fusing them, so we want to make sure it actually fuses correctly.

File: torch/csrc/jit/codegen/cuda/codegen.cpp (hunk 3 above)
Comment (csarofeen) [reply torch/csrc/jit/codegen/cuda/codegen.cpp:85]: All scalars that come in from the JIT are of type double, int64_t or bool. I wanted to reflect that design decision in our system. Follow up PR removes Float and Half completely: https://github.com/csarofeen/pytorch/pull/537

If we want to go for a full type system, we need to more carefully design scalar type promotion. Since we only fuse operations involving tensors and tensors dominate type promotion I wanted the design a bit more specific to reflect that.

File: torch/csrc/jit/codegen/cuda/codegen.cpp (hunk 4 above)
Comment (csarofeen) [reply torch/csrc/jit/codegen/cuda/codegen.cpp:89]: Reverted this in https://github.com/csarofeen/pytorch/pull/537 . All types come in from the JIT as a double, so I'm trying to reflect that.

File: torch/csrc/jit/codegen/cuda/executor_kernel_arg.h (hunk 5 above)
Comment (csarofeen) [reply torch/csrc/jit/codegen/cuda/executor_kernel_arg.h:75]: Addressed in https://github.com/csarofeen/pytorch/pull/537

File: torch/csrc/jit/codegen/cuda/type.cpp [hunk 8]
```diff
@@ -464,6 +501,22 @@ std::string stringifyThread(const ParallelType ptype) {
   return parallel_type2string(ptype);
 }
 
+std::string typePrefix(const DataType data_type) {
+  switch (data_type) {
+    case DataType::Bool:
+      return "b";
+    case DataType::Double:
+      return "d";
+    case DataType::Float:
```
Comment (shmsong) [inline torch/csrc/jit/codegen/cuda/type.cpp:510]: Is using `f` for half type intended? Is it because we do not do any internal computation on half?

File: test/cpp/jit/test_gpu.cpp [hunk 9]
```diff
@@ -2866,14 +2866,17 @@ Val* gen_jit_operand(std::pair<ValType, DataType> desc) {
   if (desc.first == ValType::TensorView) {
     return makeSymbolicTensor(2, desc.second);
   } else if (desc.first == ValType::Scalar) {
-    if (desc.second == DataType::Float)
```
Comment (shmsong) [inline test/cpp/jit/test_gpu.cpp:2869]: Do we need to cover half here?

**Review by shmsong — APPROVED** [review]
Overall looks great. Had some minor questions on details.

File: test/cpp/jit/test_gpu.cpp (hunk 9 above)
Comment (csarofeen) [reply test/cpp/jit/test_gpu.cpp:2869]: No, we don't take any Half scalars. A follow up PR completely removes half and float scalars so we always use double.

File: torch/csrc/jit/codegen/cuda/type.cpp (hunk 7 above)
Comment (csarofeen) [reply torch/csrc/jit/codegen/cuda/type.cpp]: Did this somewhere, uncertain if I put it in this PR or the follow ups.
